### PR TITLE
[AERIE-1811] Scheduling Language Resource Constraints

### PR DESCRIPTION
**Pull request by mattdailis**
* **Tickets addressed:** AERIE-1811
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

~Opening a DRAFT PR to allow people to provide feedback before I've crossed all my "i"s and dotted my "t"s.~

This PR expands the CoexistenceGoal to be able to place activities in windows defined by state constraints.

Constraints supported:
- [x] during activity type
- [x] greater than
- [x] less than
- [x] equal to
- [x] not equal to
- [x] between
- [x] composite AND
- [x] composite OR
- [x] Transition from/to values

TODO prior to merging:
- [x] Split "constraints" DSL out from scheduling DSL into a separate TS file

## Verification

Testing is done via the SchedulingIntegrationTests.

## Documentation

When this PR is ready to be merged, we definitely need to document all of these.

## Future work

- Type check which resources can be used in numeric comparisons

### CODE REVIEW

File: deployment/hasura/metadata/actions.graphql
Comment (dyst5422) [inline deployment/hasura/metadata/actions.graphql:154]: Very Smart!

File: scheduler-server/scheduling-dsl-compiler/src/libs/mission-model-generated-code.ts
Comment (dyst5422) [inline scheduler-server/scheduling-dsl-compiler/src/libs/mission-model-generated-code.ts]: I think this would be a prime candidate for an enum so that they can have a good autocompletion for resources `Resource.someResource`

File: scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts
Comment (dyst5422) [inline scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: I think the use of the kind name in a second place here is a good motivator for making the node kinds an enum. That way we have a symbol to refactor vs having to change strings all over the place

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/TypescriptCodeGenerationService.java
Comment (dyst5422) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/TypescriptCodeGenerationService.java]: See enum comment above

**Review by dyst5422 — CHANGES_REQUESTED** [review]

File: scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts
Comment (mattdailis) [reply scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: That's a good idea - playing around with it now, it seems like intellisense works better with the enum than it did with the union type, probably because of the additional context:

Screen Shot 2022-05-04 at 12 40 01 PM

Comment (dyst5422) [reply scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: Might make sense to have a super intuitive mapping to a name that doesn't require the brackets and quotes - I think its not particularly intuitive to have to do the string indexing. Also not sure if that is really a viable option the way it was for command stems.

Comment (mattdailis) [reply scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: I would think the most intuitive form of a resource name is the exact string (any mangling we do would make it less intuitive in my opinion... but maybe that's a Basak/users question).

Comment (dyst5422) [reply scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts]: I agree that the most intuitive form of the name is the exact string, but that requires a less intuitive string indexing. I think the question is which is the lesser evil.

File: scheduler-server/scheduling-dsl-compiler/src/libs/mission-model-generated-code.ts
Comment (mattdailis) [reply scheduler-server/scheduling-dsl-compiler/src/libs/mission-model-generated-code.ts]: Resource is now a string enum 👍

**Review by patkenneally — APPROVED** [review]

File: scheduler-server/scheduling-dsl-compiler/src/libs/windows-edsl-fluent-api.ts
Comment (dyst5422) [inline scheduler-server/scheduling-dsl-compiler/src/libs/windows-edsl-fluent-api.ts]: the any types here are a bit concerning. Is there anyway we can make this a type that we replace with generated code at runtime? Like we do with ActivityTypes and Resources

Comment (mattdailis) [reply scheduler-server/scheduling-dsl-compiler/src/libs/windows-edsl-fluent-api.ts]: Latest commit attempts to address this problem using function overloading.

I don't want to delve too deep into it in this PR, since @JoelCourtney's work parallels this. If we decide this is a good way to do things, hopefully we can reuse some of these ideas on the Constraints side.

Comment (mattdailis) [reply scheduler-server/scheduling-dsl-compiler/src/libs/windows-edsl-fluent-api.ts]: After a discussion with @dyst5422 and @adrienmaillard , I've updated this to use [Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html) instead of function overloading. Let me know what you think!

I'll squash the last two commits before merging this PR, since one of them uses function overloading, and the other uses conditional types.

I'll echo the sentiment in my previous comment that this code will need to move out of the scheduler and into the Constraints eDSL during the next sprint.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java
Comment (adrienmaillard) [inline scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java]: Why not an accessor?

**Review by adrienmaillard — APPROVED** [review]
Nothing blocking for me.

File: scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java
Comment (mattdailis) [reply scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java]: Updated to use an accessor :+1:

https://github.com/NASA-AMMOS/aerie/blob/4fffd68931d5a1160ae6abcefeae068de2258711/scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java#L130-L132

**Review by dyst5422 — APPROVED** [review]
Matt and I had a quick Slack chat about how it might make sense to replace the and/or infix operators with Goal.Any([]) and Goal.All([]) collection operators, but I'll leave it up to him whether to implement that in this PR or not

**Comment by mattdailis** [comment]
Last force push renames a number of things to be more similar to the constraints verbage:
- `gt` became `greaterThan`
- `lt` became `lessThan`
- `eq` became `equalTo`
- `neq` become `notEqualTo`
- `and` became `all`, and was converted to a static method
- `or` became `any` and was converted to a static method

I'll do one last survey to make sure things like `>=` and `<=` are expressible, then I'll merge.

EDIT: I suppose for the time being, it'll need to be:
```typescript
WindowSet.any(
  WindowSet.lessThan(...),
  WindowSet.equalTo(...)
)
```

Another note: discrete vs linear resources are not properly differentiated in this PR. That will come via the constraints language in an upcoming PR.

**Pull request merged**
